x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -919,11 +923,26 @@ public boolean tryRead(ByteBuffer buf, long offset) throws IOException {
             if (startRegion != endRegion) {
                 return false;
             }
-            final CacheFileRegion fileRegion = get(cacheKey, length, startRegion);
-            if (fileRegion.tracker.checkAvailable(getRegionRelativePosition(end)) == false) {
+            long now = threadPool.relativeTimeInMillis();
+            final CacheFileRegion fileRegion;
+            var r = lastAccessedRegion;
+            if (r != null && r.regionKey.region == startRegion && now - lastAccessTime <= minTimeDelta) {
+                fileRegion = r;
+            } else {
+                fileRegion = get(cacheKey, length, startRegion);
+            }
+            if (fileRegion.tracker.checkAvailable(end - getRegionStart(startRegion)) == false) {
```

Comment:
This was buggy before for end falling directly on the end of a region the relative offset came out to `0` and the available check always returns true.